### ISSUE TITLE
Composer gets stuck in a loop when getting 404 from GitLab private repo due to invalid credentials

### ISSUE THREAD

**Issue by Jean85** [composer/composer#10353]
While checking for #10352, I caught a strange bug that creates a loop during `composer install` with a private GitLab project as a dependency and credentials with insufficient permissions (in my case I was trying to use GitLab CI's `$CI_JOB_TOKEN`).
Data is partially redacted and reduced since it's from a proprietary project.

My `composer.json`:

```json
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://gitlab.example.com/foo/bar.git"
        }
    ],
    "require": {
        "foo/bar": "^1.0"
    },
    "config": {
        "preferred-install": {
            "*": "dist"
        },
        "gitlab-domains": [
            "gitlab.example.com"
        ]
    }
}
```

Reproduced using Composer 2.1.8, 2.1.14 (latest stable), 2.2.0-RC1 (latest preview)

When I run this command:
`composer config gitlab-token.gitlab.facile.it gitlab-ci-token $(CI_JOB_TOKEN)`
(as suggested by https://stackoverflow.com/a/50163888) and followed by
`composer install -vvv`

I get the following output:

```
Running 2.2.0-RC1 (2021-12-08 15:17:19) with PHP 7.4.23 on Linux / 5.4.0-74-generic
Reading ./composer.json (/var/www/foo/test/composer.json)
Loading config file ./composer.json (/var/www/foo/test/composer.json)
Loading config file /var/www/foo/test/auth.json
Reading /var/www/foo/test/auth.json
Checked CA file /etc/pki/tls/certs/ca-bundle.crt does not exist or it is not a file.
Checked directory /etc/pki/tls/certs/ca-bundle.crt does not exist or it is not a directory.
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/var/www/foo/test): git branch -a --no-color --no-abbrev -v
Executing command (/var/www/foo/test): git rev-list remotes/origin/master..98968a35fd7b371828b2ada3705b50003683846f
Executing command (/var/www/foo/test): git rev-list remotes/origin/develop..98968a35fd7b371828b2ada3705b50003683846f
Executing command (/var/www/foo/test): git describe --exact-match --tags
Failed to initialize global composer: Composer could not find the config file: /home/foo/.composer/composer.json
Reading ./composer.lock (/var/www/foo/test/composer.lock)
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Reading ./composer.lock (/var/www/foo/test/composer.lock)
Generating rules
Resolving dependencies through SAT
Looking at all rules.
Dependency resolution completed in 0.003 seconds
Package operations: 214 installs, 0 updates, 0 removals

[REDACTED]

  - Downloading composer/package-versions-deprecated (1.11.99.2)
Downloading https://api.github.com/repos/composer/package-versions-deprecated/zipball/c6522afe5540d5fc46675043d3ed5a45a740b27c
  - Downloading symfony/flex (v1.13.4)

[...]

  - Downloading wrep/daemonizable-command (dev-symfony5-support 30eb0eb)
[302] https://api.github.com/repos/php-fig/event-dispatcher/zipball/dbefd12671e8a14ec7f180cab83036ed26714bb0
Following redirect (1) https://codeload.github.com/php-fig/event-dispatcher/legacy.zip/dbefd12671e8a14ec7f180cab83036ed26714bb0
[302] https://api.github.com/repos/php-fig/container/zipball/8622567409010282b7aeebe4bb841fe98b58dcaf
Following redirect (1) https://codeload.github.com/php-fig/container/legacy.zip/8622567409010282b7aeebe4bb841fe98b58dcaf

[...]

[200] https://codeload.github.com/rectorphp/rector/legacy.zip/54fd7d08726cdf5b71e3b65a98ae82935fb297b3
Writing /home/foo/.composer/cache/files/rector/rector/6f037c51b90c4f73a8b50445a7fdea6467443d6b.zip into cache from /var/www/foo/test/vendor/composer/tmp-8d95b85d0b9bbc18339d505aa59f822e
Using HTTP basic authentication with username "gitlab-ci-token"
Downloading https://gitlab.example.com/api/v4/projects/foo%2Fbar/repository/archive.zip?sha=6d67bfa34d00305880984532757d6667fee2f4ad
[404] https://gitlab.example.com/api/v4/projects/foo%2Fbar/repository/archive.zip?sha=6d67bfa34d00305880984532757d6667fee2f4ad
Executing command (CWD): git config gitlab.accesstoken
Executing command (CWD): git config gitlab.deploytoken.user
Downloading https://gitlab.example.com/api/v4/projects/foo%2Fbar/repository/archive.zip?sha=6d67bfa34d00305880984532757d6667fee2f4ad
[404] https://gitlab.example.com/api/v4/projects/foo%2Fbar/repository/archive.zip?sha=6d67bfa34d00305880984532757d6667fee2f4ad
Executing command (CWD): git config gitlab.accesstoken
Executing command (CWD): git config gitlab.deploytoken.user
```
And then it goes on and on in a loop due to the 404.
The URL that Composer generates is totally valid, if I run it in my browser it goes straight to download the archive.

**Comment by Seldaek** [comment]
See https://github.com/composer/composer/blob/f051a42bc965c12b1f5ca7ba0aefe205d559377b/src/Composer/Util/AuthHelper.php#L122-L137 and specifically line 129 should prevent this endless looping. But it seems like your case gitlab-ci-token was the username and not the password. Maybe that's the bug, gitlab-ci-token should be checked against `$auth['username']`? I can't repro this easily so if you want it fixed please try and make a PR and experiment.

**Comment by fonsecas72** [comment]
I see this behavior too. 

Seems simple to replicate. 

```
composer config gitlab-token.gitlab.com gitlab-ci-token random-string
composer install -vvv
```
(actually passing --ignore-platform-reqs but i guess it does not matter )

Have you sorted it @Jean85 ?

**Comment by Seldaek** [comment]
Fixed by #10748
